druid-shell/src/mac/mod.rs

```diff
@@ -235,14 +271,39 @@ extern "C" fn draw_rect(this: &mut Object, _: Sel, dirtyRect: NSRect) {
         let view_state: *mut c_void = *this.get_ivar("viewState");
         let view_state = &mut *(view_state as *mut ViewState);
         let anim = (*view_state).handler.paint(&mut piet_ctx);
+        piet_ctx.finish();
+        // TODO: log errors
 
-        // TODO: handle animation
+        if anim {
+            // TODO: synchronize with screen refresh rate using CVDisplayLink instead.
+            let () = msg_send!(this as *const _, performSelectorOnMainThread: sel!(redraw)
+                withObject: nil waitUntilDone: NO);
+        }
 
         let superclass = msg_send![this, superclass];
         let () = msg_send![super(this, superclass), drawRect: dirtyRect];
     }
 }
 
+extern "C" fn run_idle(this: &mut Object, _: Sel) {
+    unsafe {
+        let view_state: *mut c_void = *this.get_ivar("viewState");
```

Comment:
Do we need to do any null checking or pointer validation?

Reply:
No. I consider `viewState` ivar to basically be a Rust object (a Box) owned by the NSView. There is a much more subtle case later in `add_idle`, where the nsview pointer loaded might be nil, but in that case the default behavior of msg_send is valid. I should add a comment there.